run/markdown-preview/editor/main.js

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const express = require('express');
+const handlebars = require('handlebars');
+const { readFileSync } = require('fs');
+const renderRequest = require('./render.js');
+
+const app = express();
+app.use(express.json());
+
+let url, isAuthenticated, markdownDefault, compiledTemplate, template;
+
+const init = () => {
+  url = process.env.EDITOR_UPSTREAM_RENDER_URL;
+  if (!url) throw Error ("No configuration for upstream render service: add EDITOR_UPSTREAM_RENDER_URL environment variable");
+  isAuthenticated = !process.env.EDITOR_UPSTREAM_UNAUTHENTICATED;
+  if (!isAuthenticated) console.log("Editor: starting in unauthenticated upstream mode");
+  return {url, isAuthenticated};
+};
+
+const service = init();
+
+// Load the template files and serve them with the Editor service.
+const buildTemplate = () => {
+  try {
+    markdownDefault = readFileSync(__dirname + '/templates/markdown.md');
+    const indexTemplate = handlebars.compile(readFileSync(__dirname + '/templates/index.html', 'utf8'));
+    compiledTemplate = indexTemplate({default: markdownDefault});
+    return compiledTemplate;
+  } catch(err) {
+    throw Error ('Error loading template: ', err);
+  }
+};
+
+app.get('/', (req, res) => { 
+  try {
+    template = buildTemplate();
```

Comment:
Because the file load and compilation of the template has a good amount of overhead, we only want to compile indexTemplate once, load markdownDefault once, but do want to render the template on demand.  It's a bit contrived since we don't have a parameter that varies by request: maybe adding one would be useful? In the meantime to illustrate what I mean:

```suggestion
    if (!template) template = buildTemplate();
```

Reply:
Understood, and updated. Using 'template' as the conditional variable to determine whether to run buildTemplate() makes the most sense to me, as opposed to creating an arbitrary parameter to determine whether the template should be rendered again.